[CAS] Teach swift compiler to compute cache key and store outputs into CAS

cachemeifyoucan opened the pull request:
Teach swift compiler how to compute cache key and store outputs into CAS. It contains:

* Cache key computation logic
* CAS output backend that mirrors the output into CAS
* Add cache lookup entry for outputs in ActionCache
* A swift-driver tool that can be used to inspect caching functions.

cachemeifyoucan committed:
Teach swift to compute cache key for compiler outputs

Teach swift how to serialize its input into CAS to create a cache key
for compiler outputs. To compute the cache key for the output, it first
needs to compute a base-key for the compiler invocation. The base key is
computed from: swift compiler version and the command-line arguments for
the invocation.

Each compiler output from swift will gets its own key. The key for the
output is computed from: the base key for the compiler invocation + the
primary input for the output + the output type.

cachemeifyoucan committed:
[CAS] Add test case for cache key computation

Add tool swift-cache-tool for caching related testing and inspection.

cachemeifyoucan committed:
Mirror outputs into CAS using CASOutputBackend

When `-enable-cas` option is used, mirror the output into CAS so it can
be fetched from CAS with the cache key in the future.

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test macOS platform

benlangmuir commented on test/CAS/cache_key_compute.swift:
```diff
@@ -0,0 +1,43 @@
+// RUN: mkdir -p %t/cas
+
+/// Doesn't run because the command doesn't have CAS enabled.
+// RUN: not %cache-tool -cas-path %t/cas -cache-tool-action print-base-key -- %target-swift-frontend %s -c
```
should we check the error message?

benlangmuir commented on lib/Frontend/CompileJobCacheKey.cpp:
```diff
@@ -0,0 +1,81 @@
+
+  llvm::cas::HierarchicalTreeBuilder Builder;
+  Builder.push(
+      llvm::cantFail(CAS.storeFromString(None, CommandLine)),
```
This should report the Error. Same with the other cantFail below.

benlangmuir commented on lib/Frontend/CompileJobCacheKey.cpp:
```diff
@@ -0,0 +1,81 @@
+using namespace swift;
+
+// TODO: Rewrite this into CASNodeSchema.
+llvm::Expected<llvm::cas::ObjectRef> swift::createCompileJobBaseCacheKey(
```
How are input file contents accounted for here?

cachemeifyoucan replied:
It is expected to come from `-cas-fs` argument. Currently, swift-frontend will produce the cache key even when `-cas-fs` is not used. This is risky since file content are in the cache key and the cache entry can't really be used but on the other hand, it makes writing unit test easier in swift repo since compute `cas-fs` for a swift invocation is not easy.

benlangmuir replied:
Makes sense. How about adding a warning if there is no cas-fs?  Would be good to explain this in a comment here too.  I'm fine with tests cheating here, just don't want to accidentally use this in production 😸

cachemeifyoucan replied:
True. Let me actually just add a hidden flag like `-allow-unstable-cache-key-for-test` which we can use in test, otherwise, it will just error out if caching is used without `-cas-fs`

benlangmuir commented on lib/DriverTool/swift_cache_tool_main.cpp:
```diff
@@ -0,0 +1,336 @@
+      Table->printHelp(llvm::outs(), ExecutableName.c_str(),
+                       "Swift Cache Tool", options::SwiftCacheToolOption, 0,
+                       /*ShowAllAliases*/ false);
+      return 1;
```
Help should probably return 0.

cachemeifyoucan committed:
[Basic] Move SupplementaryOutputPaths declaration to .def

Move all the declarations for SupplementaryOutputPaths to a def file to
keep variables and function interfaces in sync.

cachemeifyoucan commented:
@swift-ci please smoke test

benlangmuir reviewed (approved):
LGTM, but please wait for at least one more reviewer to approve since I'm not that familiar with the swift frontend.

cachemeifyoucan commented:
@swift-ci please smoke test

artemcm commented on include/swift/Frontend/CachingUtils.h:
```diff
@@ -0,0 +1,50 @@
+                                      llvm::cas::ActionCache &Cache,
+                                      StringRef Path, StringRef Bytes,
+                                      llvm::cas::ObjectRef BaseKey,
+                                      StringRef CorrespondingInput,
```
What about many-to-one cases, like WMO?

artemcm commented on lib/Frontend/CompileJobCacheKey.cpp:
```diff
@@ -0,0 +1,83 @@
+
+  // TODO: Improve this list.
+  static const std::vector<std::string> removeArgAndNext = {
+      "-o", "-supplementary-output-file-map", "-serialize-diagnostics-path",
```
Much like we have `DoesNotAffectIncrementalBuild`, we can add a `DoesNotContributeToCASKey` flag to argument definitions, in the future.

cachemeifyoucan replied:
Sounds like a good idea but we actually would love to have argument canonicalization so we can serialize/deserialize CompilerInvocation for caching purpose.

artemcm commented on lib/Frontend/CompileJobCacheKey.cpp:
```diff
@@ -0,0 +1,83 @@
+  // Don't count the `-frontend` in the first location since only frontend
+  // invocation can have a cache key.
+  if (Args.size() > 1 && StringRef(Args.front()) == "-frontend")
+    Args = Args.drop_front();
```
Is this not mutating the same underlying storage as was passed to `swift::performFrontend`?

artemcm commented on lib/Frontend/Frontend.cpp:
```diff
@@ -452,10 +470,10 @@ void CompilerInstance::setupOutputBackend() {
 
 bool CompilerInstance::setup(const CompilerInvocation &Invoke,
-                             std::string &Error) {
+                             std::string &Error, ArrayRef<const char *> Args) {
```
Can this be made work with `ParsedArgs`? 
Previously, downstream of  `parseArgs`, the rest of the compiler does not consult the raw argument values, for several good reasons (e.g. de-duplication). 
I think there's value to keeping that convention.

cachemeifyoucan replied:
Maybe ParsedArgs is better, but it has to be render backed into raw argument value for cache key currently. ParsedArg doesn't de-duplicate or handle ordering difference. The ideal solution is doing a canonicalization for CompilerInvocation

artemcm replied:
Okay then, thanks, let's leave it as-is for now, but these details are adding up to make canonicalization increasingly important. 👍🏼

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test
